python/mpt/src/mpt/trie_diff.cairo

```diff
@@ -289,19 +447,924 @@ func _process_storage_diff{
+            let full_path = nibble_list_to_bytes(updated_path);
+            let full_path_b32 = Bytes_to_Bytes32(full_path);
+
+            let opt_left_leaf = OptionalLeafNode(l_leaf.value);
+            let opt_right_leaf = OptionalLeafNode(r_leaf.value);
+
+            if (account_address.value != 0) {
+                _process_account_diff(path=full_path_b32, left=opt_left_leaf, right=opt_right_leaf);
+                return ();
+            }
+
+            _process_storage_diff(
+                address=account_address,
+                path=full_path_b32,
+                left=opt_left_leaf,
+                right=opt_right_leaf,
+            );
+
+            return ();
+        }
+
+        // Different path -> deleted old leaf, create new leaf
+        let updated_left_path = Bytes__add__(path, l_leaf.value.rest_of_key);
+        let updated_left_path_bytes = nibble_list_to_bytes(updated_left_path);
+        let updated_left_path_b32 = Bytes_to_Bytes32(updated_left_path_bytes);
+
+        let updated_right_path = Bytes__add__(path, r_leaf.value.rest_of_key);
+        let updated_right_path_nibbles = nibble_list_to_bytes(updated_right_path);
+        let updated_right_path_b32 = Bytes_to_Bytes32(updated_right_path_nibbles);
+
+        let leaf_null = OptionalLeafNode(cast(0, LeafNodeStruct*));
+        let opt_right_leaf = OptionalLeafNode(r_leaf.value);
+        let opt_left_leaf = OptionalLeafNode(l_leaf.value);
+
+        if (account_address.value != 0) {
+            _process_account_diff(path=updated_left_path_b32, left=opt_left_leaf, right=leaf_null);
+            _process_account_diff(
+                path=updated_right_path_b32, left=leaf_null, right=opt_right_leaf
+            );
+            return ();
+        }
+
+        _process_storage_diff(
+            address=account_address, path=updated_left_path_b32, left=opt_left_leaf, right=leaf_null
+        );
+        _process_storage_diff(
+            address=account_address,
+            path=updated_right_path_b32,
+            left=leaf_null,
+            right=opt_right_leaf,
+        );
+
+        return ();
+    }
+
+    // (LeafNode(), ExtensionNode()) -> Explore the extension node's subtree for any new leaves,
+    // comparing it to the old leaf with the same key
+    if (cast(right.value.extension_node.value, felt) != 0) {
+        // remove the right node's key segment from the left leaf node
+        let r_extension = right.value.extension_node;
+        let updated_path = Bytes__add__(path, r_extension.value.key_segment);
+        %{ logger.debug_cairo("updated_path: %s", updated_path) %}
+        let r_subnode = OptionalUnionInternalNodeExtendedImpl.from_extended(
+            r_extension.value.subnode
+        );
+        // Update the left leaf node's rest of key to remove the right node's key segment
+        // TODO(unsure): verify whether this is correct or if we need to use a prefix check.
```

Comment:
I wonder if it's possible that at a certain path:
1/ leaf A is deleted (with long rest_of_key)
2/ two leaves B & C. B & C share all their rest_of_key but one nibble, but nothing with A, then there is a creation of extension + branch + 2 leaves, and the created extension doesn't share keysegment wth leaf A

but not sure about the impact

Reply:
I think we need a prefix check

Reply:
see https://github.com/kkrt-labs/keth/pull/1146
https://github.com/kkrt-labs/keth/pull/1124/files#diff-1df4a7f869c97d4b35d348ad212fe31eceff4eb9cde0ace0a95f193be50db339